sdc11073/pysoap/msgreader.py

```diff
@@ -0,0 +1,260 @@
+        all_mds = md_description_node.findall(self._pm_names.Mds)
+        for mds_node in all_mds:
+            mds = self._mk_descriptor_container_from_node(mds_node, None)
+            descriptions.append(mds)
+            add_children(mds_node)
+        return descriptions
+
+    def _read_md_state_node(self, md_state_node):
+        """
+        Parses a GetMdStateResponse or the MdState part of GetMdibResponse
+        :param md_state_node: A MdState node
+        :return: a list of state containers
+        """
+        state_containers = []
+        all_state_nodes = md_state_node.findall(self._pm_names.State)
+        for state_node in all_state_nodes:
+            try:
+                state_containers.append(self._mk_state_container_from_node(state_node))
+            except MdibStructureError as ex:
+                self._logger.error('{}_read_md_state_node: cannot create: {}', self._log_prefix, ex)
+        return state_containers
+
+    def _mk_descriptor_container_from_node(self, node, parent_handle):
+        """
+        :param node: a descriptor node
+        :param parent_handle: the handle of the parent
+        :return: a DescriptorContainer object representing the content of node
+        """
+        node_type = node.get(QN_TYPE)
+        if node_type is not None:
+            node_type = text_to_qname(node_type, node.nsmap)
+        else:
+            node_type = etree_.QName(node.tag)
+        descr_cls = self.get_descriptor_container_class(node_type)
+        return descr_cls.from_node(node, parent_handle)
+
+    def _mk_state_container_from_node(self, node, forced_type=None):
+        """
+        :param node: an etree node
+        :param forced_type: if given, the QName that shall be used for class instantiation instead of the data in node
+        """
+        if forced_type is not None:
+            node_type = forced_type
+        else:
+            node_type = node.get(QN_TYPE)
+            if node_type is not None:
+                node_type = text_to_qname(node_type, node.nsmap)
+
+        descriptor_container = None
+        st_cls = self.get_state_container_class(node_type)
+        if st_cls is None:
+            raise ValueError(f'body type {node_type} is not known')
+
+        if node.tag != self._pm_names.State:
+            node = copy.copy(node)  # make a copy, do not modify the original report
+            node.tag = self._pm_names.State
+        state = st_cls(descriptor_container)
+        state.update_from_node(node)
+        state.node = node
+        return state
+
+    def _validate_node(self, node):
+        if self._validate:
+            validate_node(node, self._xml_schema, self._logger)
+
+    @staticmethod
+    def read_wsdl(wsdl_string: str) -> etree_.ElementTree:
+        """ make am ElementTree instance"""
+        return etree_.fromstring(wsdl_string, parser=etree_.ETCompatXMLParser(resolve_entities=False))
```

Comment:
I tried to connect to a consumer of version 2 devel branch. This line leads to an error for me:
```
  File "/home/kischy/comm_module_pi/consumer.py", line 65, in run_ref_test
    client.start_all()
  File "/home/kischy/.local/lib/python3.10/site-packages/sdc11073/sdcclient/sdcclientimpl.py", line 408, in start_all
    self._mk_hosted_services(self.host_description)
  File "/home/kischy/.local/lib/python3.10/site-packages/sdc11073/sdcclient/sdcclientimpl.py", line 573, in _mk_hosted_services
    h_descr.read_metadata(soap_client)
  File "/home/kischy/.local/lib/python3.10/site-packages/sdc11073/sdcclient/sdcclientimpl.py", line 70, in read_metadata
    self._read_wsdl(soap_client, self.meta_data.wsdl_location)
  File "/home/kischy/.local/lib/python3.10/site-packages/sdc11073/sdcclient/sdcclientimpl.py", line 78, in _read_wsdl
    self.wsdl_node = self._msg_reader.read_wsdl(self.wsdl_string)
  File "/home/kischy/.local/lib/python3.10/site-packages/sdc11073/pysoap/msgreader.py", line 260, in read_wsdl
    return etree_.fromstring(wsdl_string, parser=etree_.ETCompatXMLParser(resolve_entities=False))
  File "src/lxml/etree.pyx", line 3257, in lxml.etree.fromstring
  File "src/lxml/parser.pxi", line 1911, in lxml.etree._parseMemoryDocument
ValueError: Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.
```
I changed this line to the following to fix the error:

```suggestion
    def read_wsdl(wsdl_string: str) -> etree_.ElementTree:
        """ make am ElementTree instance"""
        return etree_.fromstring(bytes(wsdl_string, encoding='utf8'), parser=etree_.ETCompatXMLParser(resolve_entities=False))
```
Maybe there is a better way to avoid this error

Reply:
To clarify: The error tells us, that is does not make sense to have an encoding for a string, that is already represented as unicode code points.
The difference between our provider and the provider in "sdc11073/examples/ReferenceTest/reference_device.py" is that ours sends `<?xml version="1.0" encoding="UTF-8"?>` at the beginning of an envelope. The provider "reference_device.py" does not do that. It might just be that our provider is wrong in sending `<?xml version="1.0" encoding="UTF-8"?>`. Do you know if it is wrong to send this kind of encoding tag?

Reply:
Sounds better than my fix above. Thank you for looking into this! I think my fix is also expensive from a computational point of view.

Reply:
perfect, ty!